File: src/dimacs.py
from sat_instance import SATInstance

def parse_cnf_file(file_name):
    sat_instance = None

    try:
        with open(file_name, 'r') as file:
            # Skip comment lines
            line = next(line for line in file if not line.startswith('c'))

            # Check for the problem line
            tokens = line.split()
            if tokens[0] != 'p' or tokens[1] != 'cnf':
                raise ValueError("Error: DIMACS file does not have proper problem line or format is not CNF")

            num_vars = int(tokens[2])
            num_clauses = int(tokens[3])
            sat_instance = SATInstance(num_vars, num_clauses)

            # Parse clauses
            clause = set()
            for line in file:
                if line.startswith('c'):
                    continue

                tokens = line.split()
                if tokens[-1] != '0':
                    raise ValueError(f"Error: clause line does not end with 0 {tokens}")

                for token in tokens[:-1]:
                    if token:
                        literal = int(token)
                        clause.add(literal)
                        sat_instance.add_variable(literal)

                sat_instance.add_clause(clause)
                clause = set()

    except FileNotFoundError:
        raise FileNotFoundError(f"Error: DIMACS file is not found {file_name}")

    return sat_instance

if __name__ == "__main__":
    sat_instance = parse_cnf_file("toy_simple.cnf")
    print(sat_instance.num_vars)
    print(sat_instance.num_clauses)
    print(sat_instance.variables)
    print(sat_instance.clauses)

File: src/dimacs.jl
include("my_types.jl")

"""
Read the DIMACS file and return the clauses and the number of variables and clauses
"""
function read_dimacs(filename::String)::Tuple{Formula, Tuple{Int, Int}}
    lines = readlines(filename)

    clauses = Formula()
    for line in lines
        # skip comments and problem line
        if startswith(line, "c") || startswith(line, "p")
            continue
        end

        # skip if line is empty
        if isempty(strip(line))
            continue
        end

        # weird stuff
        if strip(line) == "%" || strip(line) == "0"
            continue
        end
        

        # parse clause
        clause = [parse(Int, x) for x in split(strip(line)) if x != "0"]
        push!(clauses, clause)
    end

    # calculate the number of clauses
    num_clauses = length(clauses)

    # calculate the number of variables
    flat = collect(Iterators.flatten(clauses))
    num_vars = maximum(abs.(flat))

    return clauses, (num_vars, num_clauses)
end

function initialize_watched_literals(clauses::Formula)::WatchedLiterals
    watched_literals = WatchedLiterals(Dict(), zeros(Literal, length(clauses), 2))

    for (i, clause) in enumerate(clauses)
        # choose the first two literals to watch
        l1, l2 = clause[1], clause[2]

        # add the literals to the watchlists
        push!(get!(watched_literals.watchlists, l1, Int[]), i)
        push!(get!(watched_literals.watchlists, l2, Int[]), i)

        # add the literals to the warrays
        watched_literals.warray[i, :] = [l1, l2]
    end

    # make sure we don't have any 0s remaining in the warray
    @assert all(x -> x != 0, watched_literals.warray)

    return watched_literals
end


if abspath(PROGRAM_FILE) == @__FILE__
    clauses, (num_vars, num_clauses) = read_dimacs("inputs/toy_simple.cnf")
    println("Number of variables: ", num_vars)
    println("Number of clauses: ", num_clauses)
    println("Clauses: ", clauses)
end